File: 5.Strings/C5/C5.cpp
#include <bits/stdc++.h>
using namespace std;

vector<int> zFunction(const string &S) {
    int n = S.size();
    vector<int> res(n);
    int l = 0, r = 0;
    for (int i = 1; i < n; ++i) {
        if (i > r) {
            int x = 0;
            while (x + i < n && S[x] == S[x + i]) {
                x++;
            }
            res[i] = x;
            if (x > 0) {
                l = i;
                r = i + x - 1;
            }
        } else {
            if (res[i - l] < r - i + 1) {
                res[i] = res[i - l];
            } else {
                int x = r - i + 1;
                while (x + i < n && S[x + i] == S[x]) {
                    x++;
                }
                l = i;
                r = i + x - 1;
                res[i] = x;
            }
        }
    }
    return res;
}

int findOccurrences(const string &s, const string &t) {
    string concat = t + "$" + s;
    vector<int> zf = zFunction(concat);
    int res = 0;
    int t_len = t.size();
    for (int i = t_len + 1; i < concat.size(); ++i) {
        if (zf[i] == t_len) {
            res++;
        }
    }
    return res;
}

int main() {
    string a, b;
    getline(cin, a);
    getline(cin, b);

    int count = 0;
    int a_len = a.size();
    int b_len = b.size();

    for (int i = 0; i <= a_len - b_len; ++i) {
        string st = a.substr(i, b_len);
        if (st != b) {
            count += findOccurrences(st + st, b);
        } else {
            count++;
        }
    }

    cout << count << endl;

    return 0;
}

File: 5.Strings/C5/C5.py
import sys
a = str(sys.stdin.readline().strip())
b = str(sys.stdin.readline().strip())
def zFunction(S):
    res = [0] * len(S)
    l = -1
    r = -1
    for i in range(1, len(S)):
        if i > r:
            x = 0
            while x + i < len(S) and S[x] == S[x + i]:
                x += 1
            res[i] = x
            if x + i - 1 > r:
                l = i
                r = x + i - 1
        else:
            if res[i - l] + i < r:
                res[i] = res[i - l]
            else:
                x = min(res[i - l], r - i)
                while x + i < len(S) and S[x + i] == S[x]:
                    x += 1
                l = i
                r = i + x - 1
                res[i] = x
    return res

def find(s, t):
    concat = t + "$" + s
    zf = zFunction(concat)
    res = 0
    for i in range(len(t) + 1, len(concat)):
        if zf[i] == len(t):
            res += 1
    return res


count = 0
for i in range(len(a) - len(b) + 1):
    st = a[i:i + len(b)]
    if st != b:
        count += find(st + st, b)
    else:
        count += 1
sys.stdout.write(str(count))



